File: algorithms/396. Rotate Function/solution.py
class Solution(object):
    def maxRotateFunction(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        n, m, f, s = len(A), 0, 0, sum(A)
        for i in xrange(n):
            m = f = f + i * A[i]
        for i in reversed(A):
            f = f + s - n * i
            m = max(m, f)
        return m

File: algorithms/396. Rotate Function/solution.cc
class Solution {
public:
    int maxRotateFunction(vector<int>& A) {
        int n = A.size(), f = 0, m = INT_MIN;
        int sum = accumulate(A.begin(), A.end(), 0);
        for ( int i = 0; i < n; ++i ) {
            f += i * A[i];
        }
        for ( int i = n - 1; i >= 0; --i ) {
            m = max(m, f);
            f += sum - n * A[i];
        }
        return n ? m : 0;
    }
};
